File: 240206/LJHmission04.c
#include <stdio.h>
#include <wiringPi.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>

#define sw1 3          // sw1는 22번에 연결
#define sw2 4          // sw2는 23번에 연결
#define sw3 5          // sw3는 24번에 연결
#define sw4 6          // sw4는 25번에 연결
#define STEP_OUTA 27   // STEP_OUTA는 16번에 연결
#define STEP_OUTB 0    // STEP_OUTB는 17번에 연결
#define STEP_OUT2A 1   // STEP_OUT2A는 18번에 연결
#define STEP_OUT2B 24  // STEP_OUT2B는 19번에 연결

void degree45();
void degree90();
void degree180();
void program_down();

int main(void){
	wiringPiSetup();

	pinMode(sw1, INPUT);
	pinMode(sw2, INPUT);
	pinMode(sw3, INPUT);
	pinMode(sw4, INPUT);

	pinMode(STEP_OUTA, OUTPUT);
	pinMode(STEP_OUTB, OUTPUT);
	pinMode(STEP_OUT2A, OUTPUT);
	pinMode(STEP_OUT2B, OUTPUT);

	digitalWrite(STEP_OUTA, 0);
	digitalWrite(STEP_OUTB, 0);
	digitalWrite(STEP_OUT2A, 0);
	digitalWrite(STEP_OUT2B, 0);

	printf("Step Motor Control Start!!\n");
	printf("SW1 : 45도\nSW2 : 90도\nSW3 : 180도\nSW4 : Exit\n");

	while(1) {
		if(digitalRead(sw1) == 0) degree45();
		else if(digitalRead(sw2) == 0) degree90();
		else if(digitalRead(sw3) == 0) degree180();
		else if(digitalRead(sw4) == 0) program_down();
	}

	return 0;
}

void degree45() {
	printf("45도 회전합니다!!\n");
	for(int i=0;i<67;i++) {
		digitalWrite(STEP_OUTA, 1);
		usleep(2000);
		digitalWrite(STEP_OUTA, 0);
		digitalWrite(STEP_OUTB, 1);
		usleep(2000);
		digitalWrite(STEP_OUTB, 0);
		digitalWrite(STEP_OUT2A, 1);
		usleep(2000);
		digitalWrite(STEP_OUT2A, 0);
		digitalWrite(STEP_OUT2B, 1);
		usleep(2000);
		digitalWrite(STEP_OUT2B, 0);
	}
}

void degree90() {
	printf("90도 회전합니다!!\n");
	for(int i=0;i<133;i++) {
		digitalWrite(STEP_OUTA, 1);
		usleep(2000);
		digitalWrite(STEP_OUTA, 0);
		digitalWrite(STEP_OUTB, 1);
		usleep(2000);
		digitalWrite(STEP_OUTB, 0);
		digitalWrite(STEP_OUT2A, 1);
		usleep(2000);
		digitalWrite(STEP_OUT2A, 0);
		digitalWrite(STEP_OUT2B, 1);
		usleep(2000);
		digitalWrite(STEP_OUT2B, 0);
	}
}

void degree180() {
	printf("180도 회전합니다!!\n");
	for(int i=0;i<268;i++) {
		digitalWrite(STEP_OUTA, 1);
		usleep(2000);
		digitalWrite(STEP_OUTA, 0);
		digitalWrite(STEP_OUTB, 1);
		usleep(2000);
		digitalWrite(STEP_OUTB, 0);
		digitalWrite(STEP_OUT2A, 1);
		usleep(2000);
		digitalWrite(STEP_OUT2A, 0);
		digitalWrite(STEP_OUT2B, 1);
		usleep(2000);
		digitalWrite(STEP_OUT2B, 0);
	}
}

void program_down() {
		printf("프로그램을 종료합니다.\n");
		exit(0);
}

File: 240206/LJHmission04.py
import RPi.GPIO as GPIO
import time

sw_pin = [22, 23, 24, 25]     # sw1은 22번, sw2는 23번, sw3은 24번, sw4는 25번에 연결
motor_pin = [16, 17, 18, 19]  # STEP_OUTA는 16번, OUTB는 17번, OUT2A는 18번, OUT2B는 19번에 연결

GPIO.setwarnings(False)

GPIO.setmode(GPIO.BCM)

GPIO.setup(sw_pin, GPIO.IN)
GPIO.setup(motor_pin, GPIO.OUT)

GPIO.output(motor_pin, GPIO.LOW)

def degree45():
	print("45도 회전합니다!!")
	for i in range(64):
		for j in motor_pin:
			GPIO.output(j, GPIO.HIGH)
			time.sleep(0.003)
			GPIO.output(j, GPIO.LOW)

def degree90():
	print("90도 회전합니다!!")
	for i in range(128):
		for j in motor_pin:
			GPIO.output(j, GPIO.HIGH)
			time.sleep(0.003)
			GPIO.output(j, GPIO.LOW)

def degree180():
	print("180도 회전합니다!!")
	for i in range(257):
		for j in motor_pin:
			GPIO.output(j, GPIO.HIGH)
			time.sleep(0.003)
			GPIO.output(j, GPIO.LOW)

def program_down():
	print("프로그램을 종료합니다.")
	exit(0)

print("Step Motor Control Start!!")
print("SW1 : 45도\nSW2 : 90도\nSW3 : 180도\nSW4 : Exit")

try:
	while True:
		if GPIO.input(sw_pin[0]) == GPIO.LOW:
			degree45()
		elif GPIO.input(sw_pin[1]) == GPIO.LOW:
			degree90()
		elif GPIO.input(sw_pin[2]) == GPIO.LOW:
			degree180()
		elif GPIO.input(sw_pin[3]) == GPIO.LOW:
			program_down()

except KeyboardInterrupt:
	pass
finally:
	GPIO.cleanup()
